### PR TITLE
Revise Symbol tutorial

### PR DESCRIPTION
**Pull request by Ishitori**
## Description ##
Revise and improve Symbol tutorial. Added section on weight tying and more links to the documentation.

### CODE REVIEW

File: docs/tutorials/basic/symbol.md [hunk 1]
```diff
@@ -71,20 +42,14 @@ To complete this tutorial, we need:
     ```
     pip install jupyter
     ```
-- GPUs - A section of this tutorial uses GPUs. If you don't have GPUs on your machine, simply
-set the variable gpu_device to mx.cpu().
+- GPUs (optional). A section of this tutorial uses GPUs. If you don't have GPUs on your machine, simply set the variable `gpu_device` to `mx.cpu()`.
```
Comment (thomelane) [inline docs/tutorials/basic/symbol.md]: Should automatically change context depending on whether a GPU is available or not.

File: docs/tutorials/basic/symbol.md [hunk 2]
```diff
@@ -125,17 +89,14 @@ net = mx.sym.FullyConnected(data=net, name='fc1', num_hidden=128)
...
+                    shape={'data':(100, 200)},
+                    node_attrs={"shape": "oval", "fixedsize": "false"})
 ```
 
-Each symbol takes a (unique) string name. NDArray and Symbol both represent
-a single tensor. *Operators* represent the computation between tensors.
-Operators take symbol (or NDArray) as inputs and might also additionally accept
-other hyperparameters such as the number of hidden neurons (*num_hidden*) or the
-activation type (*act_type*) and produce the output.
+Each symbol takes a unique string name. `NDArray` and `Symbol` both represent a single tensor. *Operators* represent the computation between tensors. Operators take `symbol` or `NDArray` as inputs and might also additionally accept other hyperparameters such as the number of hidden neurons (`num_hidden`) or the activation type (`act_type`) and produce the output.
```
Comment (thomelane) [inline docs/tutorials/basic/symbol.md]: `symbol` -> `Symbol`. Change other cases (but not non-code versions).

File: docs/tutorials/basic/symbol.md [hunk 3]
```diff
@@ -188,14 +144,9 @@ composed = net2(data2=net1, name='composed')
 composed.list_arguments()
 ```
 
-In this example, *net2* is used as a function to apply to an existing symbol *net1*,
```
Comment (thomelane) [inline docs/tutorials/basic/symbol.md]: netx -> `netx` (for 1 and 2)

File: docs/tutorials/basic/symbol.md [hunk 4]
```diff
@@ -209,63 +160,76 @@ net.list_arguments()
...
 in3a = InceptionFactoryA(prev, 64, 64, 64, 64, 96, "avg", 32, name="in3a")
-mx.viz.plot_network(symbol=in3a, shape=shape, node_attrs={"shape":"oval","fixedsize":"false"})
+mx.viz.plot_network(symbol=in3a, shape=shape,
+                    node_attrs={"shape": "oval", "fixedsize": "false"})
 ```
 
-Finally, we can obtain the whole network by chaining multiple inception
-modules. See a complete example
-[here](https://github.com/dmlc/mxnet/blob/master/example/image-classification/symbols/inception-bn.py).
+Finally, we can obtain the whole network by chaining multiple inception modules. See a [complete example](https://github.com/dmlc/mxnet/blob/master/example/image-classification/symbols/inception-bn.py).
```
Comment (thomelane) [inline docs/tutorials/basic/symbol.md]: Undo this change. Or if you wanted a larger link could include 'complete example' in the link.

File: docs/tutorials/basic/symbol.md [hunk 5]
```diff
@@ -279,65 +243,48 @@ group.list_outputs()
...
 arg_name = c.list_arguments()  # get the names of the inputs
 out_name = c.list_outputs()    # get the names of the outputs
 # infers output shape given the shape of input arguments
-arg_shape, out_shape, _ = c.infer_shape(a=(2,3), b=(2,3))
+arg_shape, out_shape, _ = c.infer_shape(a=(2, 3), b=(2, 3))
 # infers output type given the type of input arguments
 arg_type, out_type, _ = c.infer_type(a='float32', b='float32')
 {'input' : dict(zip(arg_name, arg_shape)),
- 'output' : dict(zip(out_name, out_shape))}
+'output' : dict(zip(out_name, out_shape))}
```
Comment (thomelane) [inline docs/tutorials/basic/symbol.md]: Would say the original style is better.

File: docs/tutorials/basic/symbol.md [hunk 6]
```diff
@@ -349,7 +296,8 @@ print('number of outputs = %d\nthe first output = \n%s' % (
 
 We can evaluate the same symbol on GPU with different data.
 
-**Note** In order to execute the following section on a cpu set gpu_device to mx.cpu().
+**Note** In order to execute the following section on a cpu set `gpu_device` to `mx.cpu()`.
```
Comment (thomelane) [inline docs/tutorials/basic/symbol.md]: Make automatic. I don't know how this tutorial passes in CI? Aren't all tutorials tested on CPU too?

File: docs/tutorials/basic/symbol.md [hunk 7]
```diff
@@ -359,33 +307,21 @@ ex_gpu.forward()
...
-and ```backward``` (if the gradient is needed) to get the gradient.
+For neural nets, a more commonly used pattern is [simple_bind](https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.Symbol.simple_bind), which creates all of the argument arrays for you. Then you can call `forward`, and [backward](https://mxnet.incubator.apache.org/api/python/executor/executor.html#mxnet.executor.Executor.backward) to get gradients if needed.
 
 ### Load and Save
 
-Logically symbols correspond to ndarrays. They both represent a tensor. They both
-are inputs/outputs of operators. We can either serialize a `Symbol` object by
-using `pickle`, or by using `save` and `load` methods directly as we discussed in
-[NDArray tutorial](http://mxnet.io/tutorials/basic/ndarray.html#serialize-from-to-distributed-filesystems).
+Logically symbols correspond to NDArrays. They both represent a tensor. They both are inputs/outputs of operators. We can either serialize a `Symbol` object by using `pickle`, or by using `save` and `load` methods directly as it is explained in [NDArray tutorial](http://mxnet.io/tutorials/basic/ndarray.html#serialize-from-to-distributed-filesystems).
```
Comment (thomelane) [inline docs/tutorials/basic/symbol.md]: explained in [NDArray tutorial] -> explained in [this NDArray tutorial]

File: docs/tutorials/basic/symbol.md [hunk 8]
```diff
@@ -396,25 +332,16 @@ c.tojson() == c2.tojson()
 
 ## Customized Symbol
 
-Most operators such as `mx.sym.Convolution` and `mx.sym.Reshape` are implemented
-in C++ for better performance. MXNet also allows users to write new operators
-using any front-end language such as Python. It often makes the developing and
-debugging much easier. To implement an operator in Python, refer to
-[How to create new operators](http://mxnet.io/faq/new_op.html).
+Most operators such as [mx.sym.Convolution](https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.Convolution) and [mx.sym.Reshape](https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.reshape) are implemented in C++ for better performance. MXNet also allows users to write new operators using any front-end language such as Python. It often makes the developing and debugging much easier. To implement an operator in Python, refer to [How to create new operators](http://mxnet.io/faq/new_op.html).
```
Comment (thomelane) [inline docs/tutorials/basic/symbol.md]: Why remove the in-line code rendering? (e.g. `code`)? Can still add a link to the in-line code.

File: docs/tutorials/basic/symbol.md [hunk 9]
```diff
@@ -429,18 +356,93 @@ print({'input':arg, 'output':out})
...
-data = mx.nd.ones((2,3))*2
+data = mx.nd.ones((2,3)) * 2
 ex = b.bind(ctx=mx.cpu(), args={'a':data, 'b':data})
 ex.forward()
 ex.outputs[0].asnumpy()
 ```
 
+### Weight tying
+
+You can use same principle to tie weights of different layers. In the example below two `FullyConnected` layers share same weights and biases, but process different data. Find a full example below.
```
Comment (thomelane) [inline docs/tutorials/basic/symbol.md]: Could you explain a little more about where to look or what's going on? It's quite a long code block.

Comment (thomelane) [reply docs/tutorials/basic/symbol.md]: Useful example though!

File: docs/tutorials/basic/symbol.md [hunk 10]
```diff
@@ -429,18 +356,93 @@ print({'input':arg, 'output':out})
...
+
+result = model.predict(eval_iter).asnumpy()
+print(result)
+```
+
+## Recommended Next Steps
+
+- Learn how to [use Module API to train neural network](https://mxnet.incubator.apache.org/versions/master/tutorials/basic/module.html)
+
+- Explore ways you can [load data from disk](https://mxnet.incubator.apache.org/versions/master/tutorials/basic/data.html)
```
Comment (thomelane) [inline docs/tutorials/basic/symbol.md]: Would link to a hybridization tutorial instead.

File: docs/tutorials/basic/symbol.md [hunk 11]
```diff
@@ -429,18 +356,93 @@ print({'input':arg, 'output':out})
...
+          num_epoch=1,
+          eval_metric='acc')
+
+result = model.predict(eval_iter).asnumpy()
+print(result)
+```
+
+## Recommended Next Steps
+
+- Learn how to [use Module API to train neural network](https://mxnet.incubator.apache.org/versions/master/tutorials/basic/module.html)
```
Comment (thomelane) [inline docs/tutorials/basic/symbol.md]: Add . at end of sentences.

**Comment by thomelane** [comment]
Cheers for the updates. I think the weight tying example description should go above the code block. Or even better can the code be broken into some sections, and the descriptions be inserted at the correct places. It's a long example.

**Comment by thomelane** [comment]
Yep, looks a lot better, thanks. LGTM.

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-awaiting-review]
